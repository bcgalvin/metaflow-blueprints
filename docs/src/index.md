<div class="hero-image">
   <img src="/blueprints.png" height="250">
</div>

Metaflow Blueprints provides a collection of production-ready infrastructure constructs that integrate Metaflow with event sources [supported by Argo Events](https://argoproj.github.io/argo-events/concepts/event_source/). Built on top of [cdk8s](https://cdk8s.io/docs/latest/) and [cdktf](https://developer.hashicorp.com/terraform/cdktf), Metaflow Blueprints empowers data teams to write infrastructure as code in any of the supported languages (Python, Go, TypeScript). Some of the key benefits of this toolchain include:

:::: tabs

::: tab Infrastructure as (Real) Code, not DSLs or YAML

Through the use of [jsii](https://github.com/aws/jsii), Metaflow Blueprints allows you to define your event-driven infrastructure as code using TypeScript, Python, Java, Go, or .NET. The library is written in TypeScript and compiled to other languages via jsii, ensuring consistent behavior and type safety across all supported languages.

::: code-group
<<< @/snippets/multi-lang/index.ts [TypeScript]
<<< @/snippets/multi-lang/index.py [Python]
<<< @/snippets/multi-lang/index.go [Go]

:::

::: tab IDE support

Argo Events types are imported from the openapi spec, providing rich documentation and autocompletion in addition to the typed Kubernetes objects:

![IDE Support](/ide-support.png)

:::

::: tab Fully Support Existing Tooling

- Import **any** Terraform Module / Provider or Helm Chart / CRD
- Export as HCL or [Terraform Modules](https://developer.hashicorp.com/terraform/cdktf/concepts/modules) to integrate with existing Terraform workflows
- CDK8s compiles to json/yaml Kubernetes manifests to work with any Kubernetes tooling

:::

::: tab Leverage OOP Abstractions

Allows for building powerful abstractions for Argo Events concepts like [filters](https://argoproj.github.io/argo-events/sensors/filters/intro/) and [trigger conditions](https://argoproj.github.io/argo-events/sensors/trigger-conditions/):

<<< @/snippets/oop-abs.ts [TypeScript]

...while also providing escape hatches for more complex use cases:

<<< @/snippets/escape-hatch.ts [TypeScript]

:::

::::
