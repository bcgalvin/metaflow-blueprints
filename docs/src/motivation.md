---
prev:
  text: 'Introduction'
  link: '/'
---

# Motivation

Events have been foundational to Metaflow's success story. When the [`@schedule`](https://docs.metaflow.org/production/scheduling-metaflow-flows/introduction) decorator was introduced in the 2.1.0 release in mid 2020, it provided a powerful abstraction that allowed data scientists to easily transform a prototype on their laptop to a scheduled workflow with a single decorator and cli command.

Metaflow expanded its [event-driven capabilities](https://docs.metaflow.org/production/event-triggering) a few years laterwith the addition of the `@trigger` and `@trigger_on_finish` decorators. These extend the pattern to flow orchestration, enabling teams to build modular workflows that can trigger one another. This also opened up new opportunities for data scientists to author generic reusable flows that can be used across teams and projects.

Metaflow's story with *internal* events aligns really well with the classic Rich Hickey Simple Made Easy talk. Through clean abstractions like `@schedule` and `@trigger_on_finish`, data scientists can compose sophisticated event-driven workflows without wrestling with orchestration complexity.

Yet a significant gap remains when it comes to *external* events. While Metaflow technically supports external triggers, it is up to the developer to create the Argo Events resources and wire them up to any external systems. This is neither simple nor easy, and there are no examples to follow or best practices to guide developers through the process. This makes sense; it is really hard to anticipate all of the different ways teams can configure Kubernetes, cloud infrastructure, and handle permissions for external systems.

---

The goal of Metaflow Blueprints is to address this gap by providing a set of rock-solid primitives that connect external events to Metaflow, independent of the core Metaflow repository. As mentioned in the [toolchain](/#fully-support-existing-tooling) section, users can use their preferred programming language and selectively export just the Kubernetes manifests to set up Argo Events resources or the resulting Terraform module / HCL to integrate with existing Terraform workflows.

Since we can leverage any existing Terraform provider and Argo Events provides generic event sources, we can support a wide variety of external systems. Some examples include:

| Provider | Event Source | Pattern |
|----------|--------------|---------|
| dbt Cloud | Webhook | Job completion events |
| Snowflake | Webhook | Data changes in tables/views |
| Fivetran | Webhook | Data changes in Fivetran connector |
| AWS S3 | SQS | New files in S3 |
| AWS DynamoDB | SQS | New items in DynamoDB |
| GCP PubSub | PubSub | New messages in PubSub |
|...|...|...|

Said another way, the goal of Metaflow Blueprints is to provide the necessary pencils and instructions to make it easier for the Metaflow community to draw the owl.

![How to draw an owl](/how-to-draw-an-owl.jpg)
