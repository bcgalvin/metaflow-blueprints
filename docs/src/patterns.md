---
layout: page
sidebar: false
---


<PatternBrowser />
