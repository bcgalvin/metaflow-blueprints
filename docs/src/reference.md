# API Reference <a name="API Reference" id="api-reference"></a>



## Classes <a name="Classes" id="Classes"></a>

### Hello <a name="Hello" id="metaflow-blueprints.Hello"></a>

#### Initializers <a name="Initializers" id="metaflow-blueprints.Hello.Initializer"></a>

```typescript
import { Hello } from 'metaflow-blueprints'

new Hello()
```

| **Name** | **Type** | **Description** |
| --- | --- | --- |

---

#### Methods <a name="Methods" id="Methods"></a>

| **Name** | **Description** |
| --- | --- |
| <code><a href="#metaflow-blueprints.Hello.sayHello">sayHello</a></code> | *No description.* |

---

##### `sayHello` <a name="sayHello" id="metaflow-blueprints.Hello.sayHello"></a>

```typescript
public sayHello(): string
```





