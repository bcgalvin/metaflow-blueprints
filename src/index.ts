export * as argoEvents from './imports/argoproj.io';
