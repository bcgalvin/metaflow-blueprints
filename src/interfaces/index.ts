export * from './source-spec';
export * from './sources';
export * from './types';
