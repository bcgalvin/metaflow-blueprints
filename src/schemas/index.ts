export * from './common';
export * from './sqs';
export * from './types';
