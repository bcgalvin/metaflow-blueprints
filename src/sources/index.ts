export * from './base';
export * from './sqs';
export * from './azure-queue-storage';
export * from './gcp-pubsub';
