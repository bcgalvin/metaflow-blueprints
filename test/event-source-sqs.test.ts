import * as fs from 'fs';
import * as path from 'path';
import { expect } from '@jest/globals';
import { App, Chart, ChartProps, Testing } from 'cdk8s';
import { Construct } from 'constructs';
import * as yaml from 'js-yaml';
import { SqsEventSource, SqsEventSourceProps } from '../src';

const ASSETS_DIR = path.join(__dirname, 'assets');
const EXAMPLE_YAML_PATH = path.join(
  ASSETS_DIR,
  'sqs-event-source-argo-example.yaml',
);
const OUTPUT_YAML_PATH = path.join(ASSETS_DIR, 'sqs-event-source-output.yaml');

describe('SqsEventSource', () => {
  test('synthesizes to the expected YAML manifest', () => {
    const exampleYamlContent = fs.readFileSync(EXAMPLE_YAML_PATH, 'utf8');
    const expectedManifest = yaml.load(exampleYamlContent) as Record<
      string,
      any
    >;

    const { metadata, spec } = expectedManifest;
    const { sqs } = spec;

    const [key] = Object.keys(sqs);
    const sqsSpec = sqs[key];

    const props: SqsEventSourceProps = {
      metadata: metadata,
      spec: {
        region: sqsSpec.region,
        queue: sqsSpec.queue,
        waitTimeSeconds: sqsSpec.waitTimeSeconds,
        accessKey: sqsSpec.accessKey,
        secretKey: sqsSpec.secretKey,
        jsonBody: sqsSpec.jsonBody,
        dlq: sqsSpec.dlq,
      },
    };

    const app = new App();
    const chart = new Chart(app, 'test-chart');

    new SqsEventSource(chart, key, props);

    const outputYaml = app.synthYaml();

    fs.writeFileSync(OUTPUT_YAML_PATH, outputYaml);

    const synthesizedManifest = yaml.load(outputYaml) as Record<string, any>;

    expect(synthesizedManifest).toEqual(expectedManifest);
  });
});

test('sqs-event-source', () => {
  class SqsEventSourceChart extends Chart {
    constructor(scope: Construct, id: string, props: ChartProps = {}) {
      super(scope, id, props);

      new SqsEventSource(this, 'MySqsSource', {
        metadata: { name: 'my-sqs-source' },
        spec: {
          region: 'us-west-2',
          queue: 'my-queue',
          waitTimeSeconds: 20,
          accessKey: { name: 'aws-secret', key: 'accessKey' },
          secretKey: { name: 'aws-secret', key: 'secretKey' },
          jsonBody: true,
        },
      });
    }
  }

  const app = new App();
  const chart = new SqsEventSourceChart(app, 'SqsEventSourceChart');
  app.synth();

  expect(Testing.synth(chart)).toMatchSnapshot();
});
