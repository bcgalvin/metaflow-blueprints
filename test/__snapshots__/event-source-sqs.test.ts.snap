// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sqs-event-source 1`] = `
[
  {
    "apiVersion": "argoproj.io/v1alpha1",
    "kind": "EventSource",
    "metadata": {
      "name": "my-sqs-source",
    },
    "spec": {
      "sqs": {
        "MySqsSource": {
          "accessKey": {
            "key": "accessKey",
            "name": "aws-secret",
          },
          "jsonBody": true,
          "queue": "my-queue",
          "region": "us-west-2",
          "secretKey": {
            "key": "secretKey",
            "name": "aws-secret",
          },
          "waitTimeSeconds": 20,
        },
      },
    },
  },
]
`;
